File: LeetCodeSolutions/LeetCode_0399.py
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        d = collections.defaultdict(dict)
        for idx, (a, b) in enumerate(equations):
            d[a][b] = values[idx]
            d[b][a] = 1 / values[idx]
        for k in d:
            d[k][k] = 1.0
        res = []

        def dfs(a, b):
            if a not in d or b not in d:
                return -1.0
            if a == b:
                return 1.0

        def bfs(a, b):
            if a not in d or b not in d:
                return -1.0
            if a == b:
                return 1.0
            que = collections.deque()
            visited = set([a])
            que.append([a, 1.0])
            while que:
                cur, tmp = que.popleft()
                for s in d[cur]:
                    if s == b:
                        return tmp * d[cur][s]
                    else:
                        if s not in visited:
                            visited.add(s)
                            que.append([s, tmp * d[cur][s]])
            return -1.0

        res = [bfs(a, b) for a, b in queries]
        return res

File: LeetCodeSolutions/LeetCode_0399.cpp
class Solution {
public:
    unordered_map<string, unordered_map<string, double> > graph;
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries){
        int n = equations.size();
        for(int i = 0; i < n; ++i){
            string a = equations[i][0], b = equations[i][1];
            graph[a][b] = values[i];
            graph[b][a] = 1.0 / values[i];
        }
        vector<double> ret;
        for(auto pairs: queries){
            ret.push_back(bfs(pairs[0], pairs[1]));
        }
        return ret;
    }

    double bfs(string a, string b){
        if(graph.count(a) == 0 || graph.count(b) == 0) return -1;
        if(a == b) return 1.0;
        typedef pair<string, double> PSD;
        queue<PSD> q;
        unordered_set<string> visited;
        visited.insert(a);
        q.push(make_pair(a, 1.0));
        while(!q.empty()){
            auto p = q.front();
            q.pop();
            for (auto tmp: graph[p.first]){
                if(visited.count(tmp.first) == 0){
                    double val = tmp.second * p.second;
                    graph[a][tmp.first] = val;
                    graph[tmp.first][a] = 1 / val;
                    if(tmp.first == b) return val;
                    q.push(make_pair(tmp.first, val));
                }
            }
        }
        return -1.0;

    }
};